### ISSUE TITLE
run_regularly needs to be explicitly added to Network but not MagicNetwork

### ISSUE THREAD

**Issue by thesamovar** [brian-team/brian2#1009]
I wonder if this is what we want? It's not what I expected.

**Comment by thesamovar** [comment]
It might be my rather specific setup:

```
... make network ...
run(...)
G.run_regularly(...)
run(...)
```

This didn't work, but it might work if the run_regularly was before the first run?

**Comment by thesamovar** [comment]
I guess the bug then would be that ``contained_objects`` isn't re-checked on re-running.

**Comment by mstimberg** [comment]
Ah, right, I think it is never checked for a run, it is used in `Network.add`.

**Comment by thesamovar** [comment]
I think this counts as a bug right? Not just an unexpected behaviour?

**Comment by mstimberg** [comment]
I guess so, it is logical if you know how things are implemented behind the scenes, but we don't want users to have to think about that. I could have easily made the same mistake.

**Comment by Vigneswaran-Chandrasekaran** [comment]
I made the code, which describes the issue reported.

`#importing necessary libraries
from brian2 import *
from matplotlib import pyplot as plt

#clear any pre-defined brian variables
start_scope()

#define differential equation
tau = 10*ms
diff_eqn = '''dv/dt = (1-v)/tau : 1'''

#define neuron groups
Ng1 = NeuronGroup(10, diff_eqn, threshold='v>0.5', reset='v = 0.32',method='exact')
Ng2 = NeuronGroup(10, diff_eqn, threshold='v>0.8', reset='v = 0.22',method='exact')

#define synaptic weights connecting two neuron groups 
Sy = Synapses(Ng1,Ng2)
Sy.connect(condition='i >= j')    

#define state monitor for each neuron group
M1 = StateMonitor(Ng1, 'v', record=True)
M2 = StateMonitor(Ng2, 'v', record=True)
`
I defined the components and then used magicNetwork to run,

`
#schedule particular run for neuron groups
Ng1.run_regularly('''v = 2*v''', dt=20*ms)
Ng2.run_regularly('''v = 2.5*v''', dt=30*ms)
run(100*ms)
`
When I try to plot, my model
`
plt.plot(M1.t/ms, M1.v[0], label='Neuron 1')
plt.plot(M2.t/ms, M2.v[1], label='Neuron 2')
xlabel('Time (ms)')
ylabel('mV')
plt.show()
`
![magic](https://user-images.githubusercontent.com/33456057/55416054-c2e34480-558b-11e9-9857-b5a9171e5de0.png)

However, when I use the Network object and do the same without using the magic method,
`
#define Network object and its elements
netObj = Network()
netObj.add(Ng2)
netObj.add(Ng1)
netObj.add(Sy)
netObj.add(M1)
netObj.add(M2)

#schedule particular run for neuron groups
Ng1.run_regularly('''v = 2*v''', dt=20*ms)
Ng2.run_regularly('''v = 2.5*v''', dt=30*ms)

#run the Network object
netObj.run(100*ms)

#plot the Neuronal activity for first members of each group
plt.plot(M1.t/ms, M1.v[0], label='Neuron 1')
plt.plot(M2.t/ms, M2.v[1], label='Neuron 2')
xlabel('Time (ms)')
ylabel('mV')
plt.show()
`
I get the output plot as,
![NetObj](https://user-images.githubusercontent.com/33456057/55416302-29686280-558c-11e9-821d-8aea53cfe027.png)
which shows the run_regularly() is not taking part in the Network

**Comment by mstimberg** [comment]
Hi @Vigneswaran-Chandrasekaran : this indeed shows how `run_regularly` is not taken into account if it is added after the creation of the `Network` object. Can you turn this into a test function (see our tests in `brian2/tests` and [the developer documentation](https://brian2.readthedocs.io/en/stable/developer/guidelines/testing.html#writing-tests)) that fails, i.e. an automatic way of testing whether the issue occurs instead of plotting the result and deciding from looking at the plot? Also, it would be great if you could reduce the example/test to the bare minimum, one `NeuronGroup` with its `run_regularly` operation should be enough!

PS: If you want to put nicely formatted/highlighted code into github comments, you can use
````
```Python
<-- here goes your code
```
````

**Comment by Vigneswaran-Chandrasekaran** [comment]
Thank you @mstimberg . I'll turn it into a standard test function to perform automatic testing. 

> Also, it would be great if you could reduce the example/test to the bare minimum, one NeuronGroup with its run_regularly operation should be enough!

Yes sure. :+1: 
 
>PS: If you want to put nicely formatted/highlighted code into github comments, you can use

Thank you very much :-)

**Comment by Vigneswaran-Chandrasekaran** [comment]
I've added a test function that shows the issue,
```python
from brian2 import NeuronGroup, StateMonitor, Network
from brian2 import ms, start_scope, run

def test_both_equal():
    tau = 10*ms
    diff_eqn = '''dv/dt = (1-v)/tau : 1'''
    chg_code = '''v = 2*v'''

    Ng = NeuronGroup(1,diff_eqn,method='exact')
    M1 = StateMonitor(Ng, 'v', record=True)
    netObj = Network(Ng, M1)
    Ng.run_regularly(chg_code, dt=20*ms)
    netObj.run(100*ms)

    start_scope()
    Ng = NeuronGroup(1, diff_eqn, method='exact')
    M2 = StateMonitor(Ng, 'v', record=True)
    Ng.run_regularly(chg_code, dt=20*ms)
    run(100*ms)

    assert (M1.v == M2.v).all()

if __name__ == '__main__':
    test_both_equal()
```
@mstimberg Kindly check and correct me, If I made anything wrong.
I'm looking forward to proceed further. :blush:

**Comment by mstimberg** [comment]
Ok, looking good!  So what we want to do is: when creating the `run_regularly` object, we have to check whether the corresponding `NeuronGroup` has already been added to a `Network` object or not. Can you see how this check could be done? Ideally we would like to add it automatically to the same `Network` object in that case, but I think this is not possible with our current code, and I'm not sure it is worth doing fundamental code changes for this somewhat rare edge case. What we could easily do would be to raise a warning. Maybe that's the best idea, actually? 

Another option would be to check at `Network.run` the `contained_objects` attribute of all objects and add them if they weren't included before. However, this would do an unexpected thing in yet another – even more uncommon – edge case: imagine I do (for whatever reason) the following:
```Python
group = NeuronGroup(1, 'v: 1')
run_reg = group.run_regularly('v += 1', dt=10*ms)
net = Network(group)  # automatically adds the run_regularly operation, too
net.remove(run_reg)
net.run(100*ms)
```
If we automatically add all `contained_objects` in `Network.run` we would undo the user's `net.remove`...

@thesamovar : what do you think? Is a warning good enough?

Note that we could also check in `Network.run`, but raise a warning instead of adding it automatically. This might actually be better, because with the approach based on a check in `run_regularly` we'd also  raise a warning in cases where the user *did* add the `run_regularly` operation to their network (@Vigneswaran-Chandrasekaran , do you see why?).

**Comment by Vigneswaran-Chandrasekaran** [comment]
Hi @mstimberg ,

> when creating the run_regularly object, we have to check whether the corresponding NeuronGroup has already been added to a Network object or not. Can you see how this check could be done? 

Yes, I can check, whether  run_regulary()'s NeuronGroup is added in Network object. 

> I'm not sure it is worth doing fundamental code changes for this somewhat rare edge case. What we could easily do would be to raise a warning. Maybe that's the best idea, actually?

Making changes in the base code to patch a rare issue is surely not a good choice. I agree with you and your example clearly explains the ambiguity it would bring in when we add run_regularly() as a member of the Network object. 

> Note that we could also check in Network.run, but raise a warning instead of adding it automatically. This might actually be better because with the approach based on a check in run_regularly we'd also raise a warning in cases where the user did add the run_regularly operation to their network (@Vigneswaran-Chandrasekaran , do you see why?).

Yes, I can understand. I guess, you mean, checking in run_regularly() without knowing about Network.run() present or not, would give false warning. 
For example,
```python
    Ng = NeuronGroup(1,diff_eqn,method='exact')
    M1 = StateMonitor(Ng, 'v', record=True)
    netObj = Network(Ng, M1)
    Ng.run_regularly(chg_code, dt=20*ms)
    run(100*ms)
```
would give a warning by checking NeuronGroup and run_regularly(), though it works alright. 

I have an idea, what if we do like,
Firstly, we check Network.run()'s object is involved with run_regularly() and corresponding NeuronGroup.
If yes, we can pass a flag to the object, such that
```python
class Network():
    ......
    if flag == true:
        #add explicit run_regularly() line inside, with the user's config passed to run_regularly()
        NeuronGrp.run_regularly(.....)
NetworkObj=Network(NeuronGrp)
```
Such that we're not adding it in the contained_object, but just explicitly pasting the run_regularly() line inside the class's scope, with some warning about our approach.

But still I feel that would invite unnecessary and intractabe events. Kindly say your comments.

**Comment by mstimberg** [comment]
Sorry for not replying earlier, I had to think about the best solution myself. I just discussed it with Dan, and we now agree on the best fix:

Currently, the `Network.add` method not only adds the objects that the user requests, but also all the objects stored in their `contained_objects` attribute. This is done recursively, so `contained_objects` of `contained_objects` are added as well. When `Network.run` is called, this list of objects is used to simulate the network.

We'd like to change this as follows:
* In `Network.add`, only the objects provided as the argument should  be stored to `Network.objects`. In other words, `Network.add` should ignore `contained_objects`.
* In `Network.run`, instead of just taking the list of `Network.objects`, create a new, full list of objects by going through the list and adding all `contained_objects` recursively. This should happen in an independent list, i.e. not modify the original `Network.objects` list.

This will fix the original issue, because the `run_regularly` operations will be included in the run no matter at which time its parent object was added to the network.

Let us know if you have questions about this, looking forward to a pull request with a fix :) Have a look at https://brian2.readthedocs.io/en/stable/developer/guidelines/workflow.html in case you are not familiar with our pull request workflow.

**Comment by Vigneswaran-Chandrasekaran** [comment]
@mstimberg, It looks like you found the precise origin of the issue. I read the comment thoroughly to understand completely, what the fix is. 

> Let us know if you have questions about this, looking forward to a pull request with a fix :) 

Thanks a  lot :blush: I'll start working on it and I think I don't have any questions right now, as the comment clearly articulates the fix. 

The issue, fortunately, got related to Brian's internal, which I've been trying to understand. Now I'm very happy because by fixing this issue, I would be exposed to Brian's core implementation.

**Comment by mstimberg** [comment]
Closed via #1069
